src/core_init_atmosphere/mpas_stack.F

```diff
@@ -0,0 +1,274 @@
+module mpas_stack
+
+   implicit none
+
+   private
+
+   ! Public Subroutines and Structures
+   public :: mpas_stack_is_empty
+   public :: mpas_stack_push
+   public :: mpas_stack_pop
+   public :: mpas_stack_free
+
+   public :: mpas_stack_type, mpas_stack_payload_type
+
+   type mpas_stack_payload_type
+   end type mpas_stack_payload_type
+
+   type mpas_stack_type
+      type (mpas_stack_type), pointer :: next => null()
+      class (mpas_stack_payload_type), pointer :: payload => null()
+   end type mpas_stack_type
+
+   !***********************************************************************
+   !
+   !  module mpas_stack
+   !
+   !> \brief   MPAS Stack module
+   !> \author  Miles A. Curry
+   !> \date    04/04/19
+   !> \details
+   !>
+   !> Introduction
+   !> ==============
+   !> The MPAS stack is a simple, extensible data stack data structure for use
+   !> within the MPAS atmospheric model. It functions as a wrapper around a
+   !> polymorphic data structure to provide usage in different areas.
+   !>
+   !>
+   !> Creating a Stack
+   !> ==================
+   !> The stack data structure (`type (mpas_stack_type)`) is defined by a single `next` pointer
+   !> and a pointer to a `type (mpas_stack_payload_type)`, which is defined as a empty derived type.
+   !>
+   !> To use the stack, create a derived that extends the `mpas_stack_payload_type` type.  Define your
+   !> extended derived type with members that meets your application.
+   !>
+   !> For instance:
+   !> ```
+   !> type, extends(mpas_stack_payload_type) :: my_payload_name
+   !>    ! Define the members of your type as you wish
+   !> end type my_payload_name
+   !>
+   !> class (my_payload_name) :: item1, item2
+   !> ```
+   !>
+   !> The extended mpas_stack_payload_type define type will enable it to ride along with the stack.
+   !> It will also enable you to push the same payload twice (if need-be).
```

Comment:
It's not clear how the extended type inherently allows itself to be pushed twice on the stack. Maybe here we could just mention that the stack stores *references* to payloads, and a payload can therefore be used in multiple push operations?